### ISSUE TITLE
add suport for mpm-itk

### ISSUE THREAD

**Issue by ZUNbado** [richardforth/apache2buddy#18]
Hi, is possible to add suport for mpm-itk?

[ CRITICAL ] Unable to determine whether Apache is using worker or prefork

ii  apache2-mpm-itk                    2.2.22-13+deb7u7                  amd64        multiuser MPM for Apache 2.2

Kind regards

**Comment by richardforth** [comment]
Hi,

There are no plans to support any other MPM than prefork, officially it doesn't even support event or worker modes fully.

The reason is simple, for every new MPM we have to check for there's a ton of new IF blocks and logic that has to be added to the code, which makes it unmanageable.

You're welcome to fork this or download the code and do your own thing for mpm-itk, so long as you redistribute it with the apache 2.0 license.

Kind Regards
Richard

**Comment by ZUNbado** [comment]
Hi there again, i "patched" the script to work with mpm-itk, based on http://mpm-itk.sesse.net/. I only change one line in every occurrence

-if ( $model eq "prefork") {
+if ( $model eq "prefork" || $model eq "itk" ) {

[patch.txt](https://github.com/richardforth/apache2buddy/files/714318/patch.txt)

**Comment by richardforth** [comment]
I just wanted to thank yo for taking the time to post your patch, however, I wont be patching this to work with itk for a number of reasons:

1. the patch only really adds itk to the if statements, which is not enough to cover the many complexities of the ITK MPM which apachebuddy.pl (the original script that I myself forked), was never coded up to address,, it was only designed to work with prefork, and Ive kept to that to this day.

For example in the link you yourself provided, ITK support directives that apache2buddy doesnt understand or account for, such as MaxClientsVHost and User ID assignements, apache2buddy only works with the apache or www-data users, its very inflexible,, and it would take far too much time to develop a working script that REALLY catered for MPM_ITK

2. I am not interested in developing apache2buddy beyond what has already been done. It takes up far too much of my time, and I fear that including your patch will only raise more issues relating to an MPM that I do not support on a daily basis and dont have the expertise or experience with.

3. (2) doent mean apache2buddy is being abandoned, what iIm saying is that I'm concentrating ONLY on bugfixing and maintaining the existing codebase, and not adding any more to it. Since I forked apachebuddy.pl which was just over 1,000 lines of code, apache2buddy.pl is now a whopping 2300 lines of code.

For comparison here is the original script (which is definitely  abandonware):

https://github.com/will-parsons/apachebuddy.pl/

I also wrote an extensive article on this wiki: https://github.com/richardforth/apache2buddy/wiki on the future of apache2buddy, and where I plan to take it. in essence it is where I want it to be, and I'm not adding anything more to it.

I also gave similar consideration to these issues:
https://github.com/richardforth/apache2buddy/issues/26
https://github.com/richardforth/apache2buddy/pull/36

I dont take these decisions lightly, but the same applies here, changing a few if statements doesn't actually make the script work with MPM -ITK, as soon as I add support for MPM-ITK I can guarrantee i'll me inundate with bug reports about how certain features of MPM-ITL arent working in apache2buddy, when in truth nothing works! All we would have done here is tricked apache2Buddy into Thinking everything is OK, when in truth there is no code in apache2buddy that supports MPM-ITK< is the same way that in PR36 above, someone thought that simply adding an If statement to include Fedora, therefore "makes it supported", when in truth, I've never coded this to work on Fedora, its bleeding edge and liable to break, and I don't have Fedora as a test server option on my cloud account, not to mention of course that Fedora really isn't an enterprise class server solution, so you see why I rejected that pull request.

It seems like I went off tangent, but its related,  in that your ptch does much teh same, it adds ITK to several if statements which implies support, where there is none. The intire workings of ITK are different to prefork and the calculations performed can't be made in the same way and simply arent valid for ITK.

Feel free to modify your own local copy, and you will eventually find you run into problems that need to be addressed in code, and rather extensively. 

The issue of course is, this works fine for you, as it is and its very tempting to have those minor changes merged upstream, I get that, however, this gets run on many thousands of machines worldwide, and someone somewhere  is going to have a setup that is running ITK thats not "quite" the same as yours, and apache2buddy is going to spit out some spurious data that is either false or misleading. Resulting in a bug report that I cant fix without a few thousand lines of code to cater of the nuances of the entire MPM. Or choosing not to support ITK.

Long story short, I know what I'm doing with prefork, and I know little to nothing about ITK and from what I have read on it, ITK isn't compatible with apache2buddy, And since event and worker modes also do things differently to prefork mode, they are also unsupported.

Sorry about the big wall of text  I am incredibly honoured that you want to contribute to the code, I hope that it shows I have really thought hard about it and declined the patch with good reasons.
